Client/multiplayer_sa/CMultiplayerSA_Vehicles.cpp

```diff
@@ -107,6 +114,105 @@ static void _declspec(naked) HOOK_CAEVehicleAudioEntity__Initialise()
     }
 }
 
+static void CVehicleModelInfo_Shutdown()
+{
+    CVehicleModelInfoSAInterface* mi = nullptr;
+    _asm
+    {
+        mov mi, ecx
+    }
+
+    if (!mi)
+        return;
+
+    ((void(__cdecl*)(void*))FUNC_CBaseModelInfo_Shutdown)(mi);
+
+    delete[] mi->m_dirtMaterials;
+    mi->m_dirtMaterials = nullptr;
+}
+
+static void SetDirtTextures(CVehicleModelInfoSAInterface* mi, std::uint32_t level)
+{
+    RpMaterial** materials = mi->m_numDirtMaterials > IN_PLACE_BUFFER_DIRT_SIZE ? mi->m_dirtMaterials : mi->m_staticDirtMaterials;
+    for (std::uint32_t i = 0; i < mi->m_numDirtMaterials; i++)
+        RpMaterialSetTexture(materials[i], ms_aDirtTextures[level]);
+}
+
+#define HOOKPOS_CVehicleModelInfo_SetDirtTextures 0x5D5DBB
+#define HOOKSIZE_CVehicleModelInfo_SetDirtTextures 6
+static constexpr DWORD CONTINUE_CVehicleModelInfo_SetDirtTextures = 0x5D5DE3;
+static void _declspec(naked) HOOK_CVehicleModelInfo_SetDirtTextures()
+{
+    _asm
+    {
+        pushad
+        push ebx
+        push esi
+        call SetDirtTextures
+        add esp, 8
+        popad
+
+        jmp CONTINUE_CVehicleModelInfo_SetDirtTextures
+    }
```

Comment:
can you make the function non-naked?

Reply:
Nope

Reply:
Why not?

Reply:
Why should I? I don't see any benefits right now.

1. The MultiplayerSA project is on C++14, so it makes no difference.
2. Naked hooks are more readable than non-naked ones where we mix asm with C++ (in my humble opinion).
3. Non-naked hooks with mixed asm and C++ can be problematic (see https://github.com/multitheftauto/mtasa-blue/issues/772#issuecomment-611048892).
4. I like naked hooks.

Reply:
1. We would want to move on to C++17 or C++20. Naked functions prevent us from doing so
2. You dont need to mix it with C++ at all. Naked functions just dont have the header and footer of a function. Thats all. As far as I see you are manually pushing registers either way. Naked function is unnecessary there
3. Look up
4. bruh

Reply:
I think it’s important to consider that these hooks are often inserted somewhere in the middle of the code, and a non-naked function has its own function prologue, allocation, register saving, and its own function epilogue. In the case of such hooks, there are usually unnecessary processor instructions, especially in places like rendering functions. Personally, I believe that if a function can be naked for hooking, that option should always be chosen. For now, I adhere to the current standards of the C++14 project and do not intend to make this change, at least until such a requirement comes from the staff. If it’s really important to you, feel free to open your own PR that eliminates all naked functions, and then I will have to adjust my PR to the new standards

Reply:
Allocation and register saving? No.
Only prologue and epilogue.
There are already too much naked functions. Introducing even more wont be helpful in a long run.

In this case naked function is unnecessary as the same effect can be done with regular function. For other functions this is complicated.

Reply:
> Allocation and register saving? No.
Only prologue and epilogue.

I meant the stack, my mistake.

Reply:
> 1. We would want to move on to C++17 or C++20. Naked functions prevent us from doing so

They don't. Only functions with mixed c++/asm code affected.

Reply:
> > 1. We would want to move on to C++17 or C++20. Naked functions prevent us from doing so
> 
> They don't. Only functions with mixed c++/asm code affected.

They do. C++17 changed a little bot how naked functions work. Dont remember the exact details but with C++17 naked functions are not compilable with the current mtasa configuration.

Reply:
> C++17 changed a little bit how naked functions work.

MSVC uses a new asm code generation algorithm since c++17.
The old algorithm allocates variables in the stack with using `push` opcode.
The new algorithm expects that stack was allocated in a function prolog and uses `mov [ebp-NUM], something`.
Access to unallocated stack position causes crashes. 

Solutions:
1) Remove affected code
2) Move some code into separate functions to avoid stack usage in naked functions
3) Use [__LOCAL_SIZE ](https://learn.microsoft.com/en-us/cpp/cpp/considerations-for-writing-prolog-epilog-code?view=msvc-170)to allocate stack

Only functions with mixed c++/asm code are affected. Last time i checked, the `Game SA` project was compilable with c++17.

Reply:
Mixed asm/c++ code is fine.
Hell, even c++ standard offers such capabilities.
Naked functions can only contain asm code.
No naked functions can contain both.
Time for testing them again ig